File: apps/ios/Domain/Tests/DomainTests/DiagnosticsVectors.swift
import Foundation
@testable import Domain

// P1.6b: wires the real Diagnostics.swift port into FunctionRegistry so
// diagnostics.json's vectors un-skip. Mirrors Android's
// DiagnosticsVectors.kt, including the dynamic JSON<->DetailValue
// conversion (same shape of problem as ReconcileVectors.swift's Row
// conversion, P1.6b, same session).

private func isBoolNSNumber(_ n: NSNumber) -> Bool {
    CFGetTypeID(n) == CFBooleanGetTypeID()
}

private func asDetailValue(_ any: Any) -> DetailValue {
    if any is NSNull { return .null }
    if let s = any as? String { return .str(s) }
    if let n = any as? NSNumber {
        if isBoolNSNumber(n) { return .bool(n.boolValue) }
        let d = n.doubleValue
        if d == d.rounded(.towardZero) { return .intNum(n.int64Value) }
        return .doubleNum(d)
    }
    if let arr = any as? [Any] { return .arr(arr.map(asDetailValue)) }
    if let obj = any as? [String: Any] {
        // [String: Any] has no defined order -- fine for every current
        // vector (each has <=4 keys and the comparator is key-based, not
        // order-sensitive, for redactDetail's own equality check), but see
        // Diagnostics.swift's DetailValue.obj doc comment: this is a real,
        // if currently untested, gap for a value that round-trips through
        // formatLog's order-sensitive JSON.stringify rendering.
        return .obj(obj.map { DetailEntry(key: $0.key, value: asDetailValue($0.value)) })
    }
    fatalError("unsupported DetailValue JSON value: \(any)")
}

private func detailValueToJson(_ v: DetailValue) -> Any {
    switch v {
    case .null: return NSNull()
    case .str(let s): return s
    case .intNum(let n): return n
    case .doubleNum(let n): return n
    case .bool(let b): return b
    case .arr(let items): return items.map(detailValueToJson)
    case .obj(let entries):
        var out: [String: Any] = [:]
        for e in entries { out[e.key] = detailValueToJson(e.value) }
        return out
    }
}

private func asLogEntry(_ any: Any) -> LogEntry {
    let d = any as! [String: Any]
    let detailAny = d["detail"]
    return LogEntry(
        at: (d["at"] as! NSNumber).int64Value,
        level: d["level"] as! String,
        scope: d["scope"] as! String,
        message: d["message"] as! String,
        route: d["route"] as? String,
        detail: (detailAny != nil && !(detailAny is NSNull)) ? asDetailValue(detailAny!) : nil
    )
}

/// LogEntry.route/detail are OPTIONAL TS fields, omitted from the JSON
/// entirely when absent -- same convention as ReceiptDraft.rawText (P1.5)
/// and UpiTarget's optional fields (P1.6b, this session).
private func logEntryToJson(_ e: LogEntry) -> [String: Any] {
    var out: [String: Any] = [
        "at": e.at,
        "level": e.level,
        "scope": e.scope,
        "message": e.message,
    ]
    if let route = e.route { out["route"] = route }
    if let detail = e.detail { out["detail"] = detailValueToJson(detail) }
    return out
}

func registerDiagnosticsVectors() {
    let domain = "diagnostics"

    FunctionRegistry.register(domain: domain, fn: "redactSecrets") { input in
        let d = input as! [String: Any]
        return redactSecrets(d["input"] as! String)
    }
    FunctionRegistry.register(domain: domain, fn: "redactText") { input in
        let d = input as! [String: Any]
        return redactText(d["input"] as! String)
    }
    FunctionRegistry.register(domain: domain, fn: "redactDetail") { input in
        let d = input as! [String: Any]
        return detailValueToJson(redactDetail(asDetailValue(d["input"]!)))
    }
    FunctionRegistry.register(domain: domain, fn: "makeEntry") { input in
        let d = input as! [String: Any]
        let opts = d["opts"] as? [String: Any]
        let detailAny = opts?["detail"]
        let entry = makeEntry(
            level: d["level"] as! String,
            scope: d["scope"] as! String,
            message: d["message"] as! String,
            route: opts?["route"] as? String,
            detail: (detailAny != nil && !(detailAny is NSNull)) ? asDetailValue(detailAny!) : nil,
            at: (opts?["at"] as? NSNumber)?.int64Value
        )
        return logEntryToJson(entry)
    }
    FunctionRegistry.register(domain: domain, fn: "formatLog") { input in
        let d = input as! [String: Any]
        let entries = (d["entries"] as! [Any]).map(asLogEntry)
        let contextDict = (d["context"] as? [String: Any]) ?? [:]
        // Same order caveat as asDetailValue's object case: [String: Any]
        // has no defined order. No golden vector currently exercises more
        // than one context key, so this doesn't affect correctness today.
        let context: [(key: String, value: String?)] = contextDict.map { (key: $0.key, value: $0.value as? String) }
        return formatLog(entries, context)
    }
}

File: apps/android/domain/src/test/kotlin/com/sanvya/app/domain/diagnostics/DiagnosticsVectors.kt
package com.sanvya.app.domain.diagnostics

import com.sanvya.app.domain.vectors.FunctionRegistry
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long
import kotlinx.serialization.json.longOrNull

// P1.6a: wires the real Diagnostics.kt port into FunctionRegistry so
// diagnostics.json's vectors un-skip. `detail`/redactDetail's input is a
// fully dynamic `Record<string, unknown>` in the TS source, so this
// adapter needs the same dynamic JSON<->DetailValue conversion Reconcile.kt
// needed for Row<->RowValue (P1.6a, same session).

private fun JsonElement.asDetailValue(): DetailValue = when (this) {
    is JsonNull -> DetailValue.Null
    is JsonObject -> DetailValue.Obj(LinkedHashMap(this.mapValues { (_, v) -> v.asDetailValue() }))
    is JsonArray -> DetailValue.Arr(this.map { it.asDetailValue() })
    is JsonPrimitive -> when {
        this.isString -> DetailValue.Str(this.content)
        this.booleanOrNull != null -> DetailValue.Bool(this.booleanOrNull!!)
        this.longOrNull != null -> DetailValue.IntNum(this.longOrNull!!)
        this.doubleOrNull != null -> DetailValue.DoubleNum(this.doubleOrNull!!)
        else -> DetailValue.Str(this.content)
    }
}

private fun DetailValue.toJson(): JsonElement = when (this) {
    is DetailValue.Null -> JsonNull
    is DetailValue.Str -> JsonPrimitive(value)
    is DetailValue.IntNum -> JsonPrimitive(value)
    is DetailValue.DoubleNum -> JsonPrimitive(value)
    is DetailValue.Bool -> JsonPrimitive(value)
    is DetailValue.Arr -> JsonArray(value.map { it.toJson() })
    is DetailValue.Obj -> JsonObject(value.mapValues { (_, v) -> v.toJson() })
}

private fun JsonElement.asLogEntry(): LogEntry {
    val o = jsonObject
    val detailEl = o["detail"]?.takeIf { it !is JsonNull }
    return LogEntry(
        at = o.getValue("at").jsonPrimitive.long,
        level = o.getValue("level").jsonPrimitive.content,
        scope = o.getValue("scope").jsonPrimitive.content,
        message = o.getValue("message").jsonPrimitive.content,
        route = o["route"]?.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
        detail = detailEl?.asDetailValue() as? DetailValue.Obj,
    )
}

/** LogEntry.route/detail are OPTIONAL TS fields, omitted from the JSON
 * entirely when absent -- same convention as ReceiptDraft.rawText (P1.5)
 * and UpiTarget's optional fields (P1.6a, this session). */
private fun LogEntry.toJson(): JsonElement = JsonObject(
    buildMap {
        put("at", JsonPrimitive(at))
        put("level", JsonPrimitive(level))
        put("scope", JsonPrimitive(scope))
        put("message", JsonPrimitive(message))
        if (route != null) put("route", JsonPrimitive(route))
        if (detail != null) put("detail", detail!!.toJson())
    }
)

fun registerDiagnosticsVectors() {
    val domain = "diagnostics"

    FunctionRegistry.register(domain, "redactSecrets") { input ->
        JsonPrimitive(redactSecrets(input.jsonObject.getValue("input").jsonPrimitive.content))
    }
    FunctionRegistry.register(domain, "redactText") { input ->
        JsonPrimitive(redactText(input.jsonObject.getValue("input").jsonPrimitive.content))
    }
    FunctionRegistry.register(domain, "redactDetail") { input ->
        redactDetail(input.jsonObject.getValue("input").asDetailValue()).toJson()
    }
    FunctionRegistry.register(domain, "makeEntry") { input ->
        val o = input.jsonObject
        val optsObj = o["opts"]?.takeIf { it !is JsonNull }?.jsonObject
        val detailEl = optsObj?.get("detail")?.takeIf { it !is JsonNull }
        makeEntry(
            level = o.getValue("level").jsonPrimitive.content,
            scope = o.getValue("scope").jsonPrimitive.content,
            message = o.getValue("message").jsonPrimitive.content,
            route = optsObj?.get("route")?.takeIf { it !is JsonNull }?.jsonPrimitive?.content,
            detail = detailEl?.asDetailValue() as? DetailValue.Obj,
            at = optsObj?.get("at")?.takeIf { it !is JsonNull }?.jsonPrimitive?.long,
        ).toJson()
    }
    FunctionRegistry.register(domain, "formatLog") { input ->
        val o = input.jsonObject
        val entries = o.getValue("entries").jsonArray.map { it.asLogEntry() }
        val contextObj = o["context"]?.takeIf { it !is JsonNull }?.jsonObject ?: JsonObject(emptyMap())
        val context = contextObj.mapValues { (_, v) -> v.takeIf { it !is JsonNull }?.jsonPrimitive?.content }
        JsonPrimitive(formatLog(entries, context))
    }
}
